File: CSharp/src/Runtime/Scripts/XString/hash.cs
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

using System;
using System.Security.Cryptography;
using System.Text;

namespace FrameworkEase.Universal.X.Utility
{
    public static partial class XString
    {
        public enum HashAlgorithm : byte
        {
            MD5,
            SHA1,
            SHA256
        }

        public static string Hash(this string input, HashAlgorithm type = HashAlgorithm.MD5)
        {
            if (string.IsNullOrEmpty(input)) return string.Empty;
            return Encoding.UTF8.GetBytes(input).Hash(type);
        }

        public static string Hash(this byte[] input, HashAlgorithm type = HashAlgorithm.MD5)
        {
            if (input == null) return string.Empty;
            try
            {
                byte[] hash;
                switch (type)
                {
                    case HashAlgorithm.MD5:
                        {
                            using var md5 = MD5.Create();
                            hash = md5.ComputeHash(input);
                            break;
                        }
                    case HashAlgorithm.SHA1:
                        {
                            using var sha1 = SHA1.Create();
                            hash = sha1.ComputeHash(input);
                            break;
                        }
                    case HashAlgorithm.SHA256:
                        {
                            using var sha256 = SHA256.Create();
                            hash = sha256.ComputeHash(input);
                            break;
                        }
                    default: throw new ArgumentException($"Unsupported hash type: {type}");
                }
                var sb = new StringBuilder(hash.Length * 2);
                foreach (var b in hash) sb.Append(b.ToString("x2"));
                return sb.ToString();
            }
            catch (Exception e)
            {
                XLog.Error(e);
                return string.Empty;
            }
        }
    }
}

File: Go/XString/hash.go
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

package XString

import (
	"crypto/md5"
	"crypto/sha1"
	"crypto/sha256"
	"encoding/hex"
)

type HashAlgorithm uint8

const (
	HashAlgorithmMD5 HashAlgorithm = iota
	HashAlgorithmSHA1
	HashAlgorithmSHA256
)

func Hash[T string | []byte](input T, algorithm ...HashAlgorithm) string {
	var data []byte
	switch v := any(input).(type) {
	case string:
		if v == "" {
			return ""
		}
		data = []byte(v)
	case []byte:
		if v == nil {
			return ""
		}
		data = v
	default:
		return ""
	}

	alg := HashAlgorithmMD5
	if len(algorithm) > 0 {
		alg = algorithm[0]
	}

	var hash []byte
	switch alg {
	case HashAlgorithmMD5:
		h := md5.Sum(data)
		hash = h[:]
	case HashAlgorithmSHA1:
		h := sha1.Sum(data)
		hash = h[:]
	case HashAlgorithmSHA256:
		h := sha256.Sum256(data)
		hash = h[:]
	default:
		return ""
	}

	return hex.EncodeToString(hash)
}
